src/compiler/crystal/semantic/cleanup_transformer.cr

```diff
@@ -249,6 +249,56 @@ module Crystal
       nil
     end
 
+    def transform(node : StringInterpolation)
+      # See if we can solve all the pieces to string literals.
+      # If that's the case, we can replace the entire interpolation
+      # with a single string literal.
+      pieces = node.expressions.dup
+      solve_string_interpolation_expressions(pieces)
+
+      if pieces.all?(StringLiteral)
+        string = pieces.join(&.as(StringLiteral).value)
+        string_literal = StringLiteral.new(string).at(node)
+        string_literal.type = @program.string
+        return string_literal
+      end
+
+      if expanded = node.expanded
+        return expanded.transform(self)
+      end
+      node
+    end
+
+    private def solve_string_interpolation_expressions(pieces : Array(ASTNode))
+      pieces.each_with_index do |piece, i|
+        replacement = solve_string_interpolation_expression(piece)
+        next unless replacement
+
+        pieces[i] = replacement
+      end
+    end
+
+    private def solve_string_interpolation_expression(piece : ASTNode) : StringLiteral?
+      if piece.is_a?(ExpandableNode)
```

Comment:
Can this be moved into the case expression below?

Reply:
I guess yes. It's slightly different because... what if `Path` or `StringLiteral` are expandable nodes? This check would be skipped. That's why I preferred to do it this way.

Reply:
Can `Path` or `StringLiteral` be expandable nodes? https://github.com/crystal-lang/crystal/blob/abe4d344cc1807223982a7f629563487a0faabb5/src/compiler/crystal/semantic/ast.cr#L651-L659